R/case-variables.R

```diff
@@ -0,0 +1,149 @@
+#' 
+#' @param ... a sequence of named expressions to use as cases as well as other 
+#' properties to pass about the case variable (i.e. alias, description)
+#' @param cases a list of lists with each case condition to use each must 
+#' include at least a `name` and an `expression` element.
+#' @param else_case a single list that has at least a `name` to serve as the 
+#' case when no others match (if not specified Crunch will use the system 
+#' default "No data")
+#' @param name a character to use as the name of the case variable to create
+#' @export
+makeCaseVariable <- function (..., cases, else_case, name) {
+    dots <- list(...)
+    is_expr <- function (x) inherits(x, "CrunchLogicalExpr")
+    exprs <- Filter(is_expr, dots)
+    if (!missing(cases) & length(exprs) > 0) {
+        halt("can't have case conditions both in ... as well as in the cases ",
+             "argument, please use one or the other.")
+    }
+    if (missing(cases) & length(exprs) == 0) {
+        halt("must supply case conditions in either ... or the cases ",
+             "argument, please use one or the other.")
+    }
+    if (missing(cases) & length(exprs) > 0) {
+        # save any member of dots that isn't an expression for use later.
+        dots <- Filter(Negate(is_expr), dots)
+        cases <- mapply(function (expr, name) list(name=name, expression=expr),
+                        expr=exprs, name=names(exprs),
+                        SIMPLIFY = FALSE, USE.NAMES = FALSE)
+    }
+    
+    # make ids if the cases don't have ids.
+    need_ids <- sapply(cases, function(x) is.null(x$id))
+    cases[need_ids] <- lapply(seq_along(cases[need_ids]), function(i) {
+        cases[need_ids][[i]]$id <- i
+        return(cases[need_ids][[i]])
+    })
+    
+    # NULLify numeric_value and FALSE missing if they are empty
+    cases <- lapply(cases, function(x) {
+        if(is.null(x$numeric_value)) {
+            x['numeric_value'] <- list(NULL)
+        }
+        if(is.null(x$missing)) {
+            x['missing'] <- FALSE
+        }
+        return(x)
+    })
+    cases <- lapply(cases, ensureValidCases)
+
+    if (!missing(else_case)) {
+        # assign API default values if not given
+        if(is.null(else_case$id)) {
+            else_case['id'] <- -1L
+        }
+        if(is.null(else_case$numeric_value)) {
+            else_case['numeric_value'] <- list(NULL)
+        }
+        if(is.null(else_case$missing)) {
+            else_case['missing'] <- FALSE
+        }
+        if(!is.null(else_case$expression)) {
+            halt("else_cases should not have any conditions expression")
+        }
+        if(!is.character(else_case$name)) {
+            halt("else_cases must have a (character) name")
+        }
+        if(!is.integer(else_case$id)) {
+            halt("id must be an integer")
+        }
+        cases <- append(cases, list(else_case))
+    }
+    
```

Comment:
I might be missing it, but I think you're missing the check that `id` is unique across case definitions.